DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,801 @@
+    immutable A a4 = a;     // calls 3
+    immutable A a5 = ia;    // calls 4
+}
+```
+The proposed model enables the user to define the copying from an object of any qualified type
+to an object of any qualified type: any combination of 2 between mutable, `const`, `immutable`, `shared`,
+`const shared`.
+
+The `inout` qualifier may be used for the copy constructor parameter in order to specify that mutable, `const` or `immutable` types are
+treated the same:
+
+```d
+struct A
+{
+    this(ref inout A rhs) immutable {}
+}
+
+void main()
+{
+    A r1;
+    const(A) r2;
+    immutable(A) r3;
+
+    // All call the same copy constructor because `inout` acts like a wildcard
+    immutable(A) a = r1;
+    immutable(A) b = r2;
+    immutable(A) c = r3;
+}
+```
+
+In case of partial matching, the existing overloading and implicit conversions
+apply to the argument.
+
+#### Copy constructor call vs. Blitting
+
+When a copy constructor is not defined for a struct, initializations are treated
+by copying the contents from the memory location of the right-hand side expression
+to the memory location of the left-hand side one (called "blitting"):
+
+```d
+struct A
+{
+    int[] a;
+}
+
+void main()
+{
+    A a = A([7]);
+    A b = a;                 // mempcy(&b, &a)
+    immutable A c = A([12]);
+    immutable A d = c;       // memcpy(&d, &c)
+}
+```
+
+Whenever a copy constructor is defined for a struct all implicit blitting is disabled for
+that struct:
+
+```d
+struct A
+{
+    int[] a;
+    this(ref A rhs) {}
+}
+
+void fun(immutable A) {}
+
+void main()
+{
+    immutable A a;
+    fun(a);          // error: copy constructor cannot be called with types (immutable) immutable
+}
+```
+
+#### Interaction with `alias this`
+
+There are situations in which a struct defines both an `alias this` and a copy constructor, and
+for which assignments to variables of the struct type may lead to ambiguities:
+
+```d
+struct A
+{
+    int a;
+    immutable(A) fun()
+    {
+        return immutable A(7);
+    }
+
+    alias fun this;
+
+    this(ref A another) immutable {}
+}
+
+struct B
+{
+    int a;
+    A fun()
+    {
+        return A(7);
+    }
+
+    alias fun this;
+
+    this(ref B another) immutable {}
+}
+
+void main()
+{
+    immutable A ia;
+    A a = ia;            // 1 - calls copy constructor
+
+    A bc;
+    A b = bc;            // 2 - calls B.fun
+}
+```
+
+In situations where both the copy constructor and `alias this` are suitable
+to solve the assignment (1), the copy constructor takes precedence over `alias this`
+because it is considered more specialized (the sole purpose of the copy constructor is to
+create copies). However, if no copy constructor in the overload set matches the exact
+qualified types of the source and the destination, the `alias this` is preferred (2).
+
+#### Generating copy constructors
+
+A copy constructor is generated for a `struct` S  if one of S's fields defines
+a copy constructor that S does not define.
```

Comment:
that S does not define. -> but `S` does not.

Reply:
I think that this would change the meaning of the sentence. `S` might define some copy constructors, but it is possible that one of the fields defines a specific copy constructor that `S` does not. I think it's better we leave the wording as is.